### PR TITLE
[ember-qunit] Add this type for module hooks

### PR DESCRIPTION
**Pull request by gitKrystan**
Module hooks have access to the test context via `this`.

Please fill in this template.

- [x] Use a meaningful title for the pull request. Include the name of the package modified.
- [x] Test the change in your own code. (Compile and run.)
- [x] Add or edit tests to reflect the change. (Run with `npm test`.)
- [x] Follow the advice from the [readme](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/README.md#make-a-pull-request).
- [x] Avoid [common mistakes](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/README.md#common-mistakes).
- [x] Run `npm run lint package-name` (or `tsc` if no `tslint.json` is present).

If changing an existing definition:
- [x] Provide a URL to documentation or source code which provides context for the suggested changes: https://github.com/emberjs/ember-qunit#setup-tests
- [ ] ~If this PR brings the type definitions up to date with a new version of the JS library, update the version number in the header.~
- [x] Include [tests for your changes](https://github.com/DefinitelyTyped/DefinitelyTyped#testing)
- [ ] ~If you are making substantial changes, consider adding a `tslint.json` containing `{ "extends": "dtslint/dt.json" }`. If for reason the any rule need to be disabled, disable it for that line using `// tslint:disable-next-line [ruleName]` and not for whole package so that the need for disabling can be reviewed.~

### CODE REVIEW

**Comment by danger-public** [comment]
Inspecting the JavaScript source for this package found some properties that are not in the .d.ts files.
The check for missing properties isn't always right, so take this list as advice, not a requirement.
## ember-qunit ([<kbd>unpkg</kbd>](https://unpkg.com/browse/ember-qunit@latest/))
was missing the following properties:
1. name
1. init
1. included
1. targetOptions
1. contentFor

<details>
<summary>as well as these 3 other properties...</summary>
<p>treeForVendor, treeForAddonTestSupport, setTestGenerator</p>
</details>

<p align="right">
  Generated by :no_entry_sign: <a href="https://danger.systems/js">dangerJS</a> against da9c0d58ff7691292ebf84c8151c2f747e29f0b8
</p>

File: types/ember-qunit/index.d.ts
Comment (chriskrycho) [inline types/ember-qunit/index.d.ts:163]: This is not an unreasonable way of defining these, but it definitely took me a minute to see what you were doing: supplying overloads like this is pretty unusual. A more idiomatic way of writing this interface would look like this: 

```ts
interface ModuleFn {
  (name: string, hooks?: Hooks, nested?: (hooks: NestedHooks) => void) => void;
  (name: string, nested?: (hooks: NestedHooks) => void) => void;
}

interface Module extends ModuleFn {
  only: ModuleFn;
}

export const module: Module;
```

Both ways *work*, but this makes it clearer that we're defining a shared function interface which the `module` itself *and* the module's embedded `only` type share, and defining the `module` type as including the merge of the function type and the `only` property.

**Review by chriskrycho — CHANGES_REQUESTED** [review]
Thank you for filing this (and sorry for the delayed response)! Two quick requests here:

1. I’ve left a note on a more idiomatic (and therefore easier for other maintainers to easily parse) way of implementing the actual fix.
2. The linked docs don’t show `module.only`—do you mind linking to docs showing that the namespace-style export is supported as well as the standalone import? And are there others we should also look to add (even if in a different PR)?

File: types/ember-qunit/index.d.ts
Comment (gitKrystan) [reply types/ember-qunit/index.d.ts:163]: This is borrowed from QUnit types, and I will admit I don't fully understand what it's all about.

https://github.com/DefinitelyTyped/DefinitelyTyped/blob/1b51c6dafe512ceceaf5686a77532d5521960335/types/qunit/index.d.ts#L372-L374

Comment (chriskrycho) [reply types/ember-qunit/index.d.ts:163]: That's fair! Let's go ahead and make the change here, and I'll make a note to upstream it into the QUnit types as well.

By way of explanation: this exists to provide an overload, since you can call it with only `Hooks`, or only `NestedHooks`, or `Hooks` *and* `NestedHooks`—but the exact same signature applies to `module(...)` and to `only(...)`. Doing it the way I have gets you those same benefits but is a bit clearer and more maintainable. Let me know if you have further questions!

Also, does QUnit itself have docs on the `module.only` invocation? I didn't see them on the Ember side, but I also didn't dive that far into it!

**Review by andrewbranch — CHANGES_REQUESTED** [review]
Requesting changes to echo @chriskrycho’s request, because I think @typescript-bot is now ignoring him

**Comment by chriskrycho** [comment]
@gitKrystan, anything I can do to help unblock you on this?

**Comment by jamescdavis** [comment]
Note: #47731

File: types/ember-qunit/index.d.ts
Comment (jamescdavis) [inline types/ember-qunit/index.d.ts:186]: Because of #47731, this should be moved inside `declare global {}` to augment the global `NestedHooks` interface. I'd say put it before the augmentation of the global `QUnit` interface (the same order as in qunit's types).

File: types/ember-qunit/index.d.ts
Comment (jamescdavis) [reply types/ember-qunit/index.d.ts:163]: The recent refactor I did in #47731 should make it unnecessary to include this at all here.

**Review by jamescdavis — CHANGES_REQUESTED** [review]
A couple of changes related to #47731 (which actually makes this simpler).

**Pull request closed without merging**
